src/main/java/org/opensearch/knn/index/codec/KNN9120Codec/KNN9120PerFieldKnnVectorsFormat.java

```diff
@@ -41,23 +43,31 @@ public KNN9120PerFieldKnnVectorsFormat(final Optional<MapperService> mapperServi
             knnScalarQuantizedVectorsFormatParams -> new Lucene99HnswScalarQuantizedVectorsFormat(
                 knnScalarQuantizedVectorsFormatParams.getMaxConnections(),
                 knnScalarQuantizedVectorsFormatParams.getBeamWidth(),
-                NUM_MERGE_WORKERS,
+                getMergeThreadCount(),
                 knnScalarQuantizedVectorsFormatParams.getBits(),
                 knnScalarQuantizedVectorsFormatParams.isCompressFlag(),
                 knnScalarQuantizedVectorsFormatParams.getConfidenceInterval(),
-                null
+                getMergeExecutorService()
             )
         );
     }
 
-    @Override
     /**
      * This method returns the maximum dimension allowed from KNNEngine for Lucene codec
      *
      * @param fieldName Name of the field, ignored
      * @return Maximum constant dimension set by KNNEngine
      */
+    @Override
     public int getMaxDimensions(String fieldName) {
         return KNNEngine.getMaxDimensionByEngine(KNNEngine.LUCENE);
     }
+
+    private static int getMergeThreadCount() {
+        return KNNSettings.getIndexThreadQty();
+    }
+
+    private static ExecutorService getMergeExecutorService() {
+        return Executors.newFixedThreadPool(getMergeThreadCount());
```

Comment:
This can be done later but I think it would make sense to initialize all our executors through a single provider.
We have no control over this reference from other parts of the code with local init.

Reply:
Since this executor is more of internal and will be present per segment, I was trying to avoid keeping a global reference for this. Another thing is it closely mimics Faiss behavior where we use OMP threads to parallelize the HNSW builds.